File: reference-code/src/main/java/createcommands/arguments/suggestions/SafeSuggestions.java
package createcommands.arguments.suggestions;

import dev.jorel.commandapi.CommandAPICommand;
import dev.jorel.commandapi.arguments.Argument;
import dev.jorel.commandapi.arguments.EntitySelectorArgument;
import dev.jorel.commandapi.arguments.EntityTypeArgument;
import dev.jorel.commandapi.arguments.PotionEffectArgument;
import dev.jorel.commandapi.arguments.RecipeArgument;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.Recipe;
import org.bukkit.inventory.ShapedRecipe;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static org.bukkit.Bukkit.getServer;

class SafeSuggestions {
    static {
        JavaPlugin plugin = null;
        // #region registerCustomItem
        // Create our itemstack
        ItemStack emeraldSword = new ItemStack(Material.DIAMOND_SWORD);
        ItemMeta meta = emeraldSword.getItemMeta();
        meta.setDisplayName("Emerald Sword");
        meta.setUnbreakable(true);
        emeraldSword.setItemMeta(meta);

        // Create and register our recipe
        ShapedRecipe emeraldSwordRecipe = new ShapedRecipe(new NamespacedKey(plugin, "emerald_sword"), emeraldSword);
        emeraldSwordRecipe.shape(
            "AEA",
            "AEA",
            "ABA"
        );
        emeraldSwordRecipe.setIngredient('A', Material.AIR);
        emeraldSwordRecipe.setIngredient('E', Material.EMERALD);
        emeraldSwordRecipe.setIngredient('B', Material.BLAZE_ROD);
        getServer().addRecipe(emeraldSwordRecipe);

        // Omitted, more itemstacks and recipes
        // #endregion registerCustomItem

        // #region registerCommandWithSafeSuggestions
        // Safely override with the recipe we've defined
        List<Argument<?>> arguments = new ArrayList<>();
        arguments.add(new RecipeArgument("recipe").replaceSafeSuggestions(dev.jorel.commandapi.arguments.SafeSuggestions.suggest(info ->
            new Recipe[]{emeraldSwordRecipe, /* Other recipes here */}
        )));

        // Register our command
        new CommandAPICommand("giverecipe")
            .withArguments(arguments)
            .executesPlayer((player, args) -> {
                Recipe recipe = (Recipe) args.get("recipe");
                player.getInventory().addItem(recipe.getResult());
            })
            .register();
        // #endregion registerCommandWithSafeSuggestions

        // #region createForbiddenMobsList
        EntityType[] forbiddenMobs = new EntityType[]{EntityType.ENDER_DRAGON, EntityType.WITHER};
        List<EntityType> allowedMobs = new ArrayList<>(Arrays.asList(EntityType.values()));
        allowedMobs.removeAll(Arrays.asList(forbiddenMobs)); // Now contains everything except enderd ragon and wither
        // #endregion createForbiddenMobsList

        // #region createSafeArguments
        List<Argument<?>> safeArguments = new ArrayList<>();
        safeArguments.add(new EntityTypeArgument("mob").replaceSafeSuggestions(dev.jorel.commandapi.arguments.SafeSuggestions.suggest(
            info -> {
                if (info.sender().isOp()) {
                    // All entity types
                    return EntityType.values();
                } else {
                    // Only allowedMobs
                    return allowedMobs.toArray(new EntityType[0]);
                }
            })
        ));
        // #endregion createSafeArguments

        // #region registerSpawnMobCommand
        new CommandAPICommand("spawnmob")
            .withArguments(safeArguments)
            .executesPlayer((player, args) -> {
                EntityType entityType = (EntityType) args.get("mob");
                player.getWorld().spawnEntity(player.getLocation(), entityType);
            })
            .register();
        // #endregion registerSpawnMobCommand

        // #region createSafePotionEffectArguments
        List<Argument<?>> safeArgs = new ArrayList<>();
        safeArgs.add(new EntitySelectorArgument.OnePlayer("target"));
        safeArgs.add(new PotionEffectArgument("potioneffect").replaceSafeSuggestions(dev.jorel.commandapi.arguments.SafeSuggestions.suggest(
            info -> {
                Player target = (Player) info.previousArgs().get(0);

                // Convert PotionEffect[] into PotionEffectType[]
                return target.getActivePotionEffects().stream()
                    .map(PotionEffect::getType)
                    .toArray(PotionEffectType[]::new);
            })
        ));
        // #endregion createSafePotionEffectArguments

        // #region registerRemoveEffectCommand
        new CommandAPICommand("removeeffect")
            .withArguments(safeArgs)
            .executesPlayer((player, args) -> {
                Player target = (Player) args.get("target");
                PotionEffectType potionEffect = (PotionEffectType) args.get("potioneffect");
                target.removePotionEffect(potionEffect);
            })
            .register();
        // #endregion registerRemoveEffectCommand
    }
}

File: reference-code/src/main/kotlin/createcommands/arguments/suggestions/SafeSuggestions.kt
package createcommands.arguments.suggestions

import dev.jorel.commandapi.CommandAPICommand
import dev.jorel.commandapi.arguments.Argument
import dev.jorel.commandapi.arguments.EntitySelectorArgument
import dev.jorel.commandapi.arguments.EntityTypeArgument
import dev.jorel.commandapi.arguments.PotionEffectArgument
import dev.jorel.commandapi.arguments.RecipeArgument
import dev.jorel.commandapi.arguments.SafeSuggestions
import dev.jorel.commandapi.executors.PlayerCommandExecutor
import org.bukkit.Bukkit.getServer
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.entity.EntityType
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.Recipe
import org.bukkit.inventory.ShapedRecipe
import org.bukkit.plugin.java.JavaPlugin
import org.bukkit.potion.PotionEffectType

fun safeSuggestions() {
    val plugin = object : JavaPlugin() {}
    // #region registerCustomItem
    // Create our itemstack
    val emeraldSword = ItemStack(Material.DIAMOND_SWORD)
    val meta = emeraldSword.itemMeta
    meta?.setDisplayName("Emerald Sword")
    meta?.isUnbreakable = true
    emeraldSword.itemMeta = meta

    // Create and register our recipe
    val emeraldSwordRecipe = ShapedRecipe(NamespacedKey(plugin, "emerald_sword"), emeraldSword)
    emeraldSwordRecipe.shape(
        "AEA",
        "AEA",
        "ABA"
    )
    emeraldSwordRecipe.setIngredient('A', Material.AIR)
    emeraldSwordRecipe.setIngredient('E', Material.EMERALD)
    emeraldSwordRecipe.setIngredient('B', Material.BLAZE_ROD)
    getServer().addRecipe(emeraldSwordRecipe)

    // Omitted, more itemstacks and recipes
    // #endregion registerCustomItem

    // #region registerCommandWithSafeSuggestions
    // Safely override with the recipe we've defined
    val arguments = listOf<Argument<*>>(
        RecipeArgument("recipe").replaceSafeSuggestions(SafeSuggestions.suggest {
            arrayOf(emeraldSwordRecipe /* Other recipes here */)
        })
    )

    // Register our command
    CommandAPICommand("giverecipe")
        .withArguments(arguments)
        .executesPlayer(PlayerCommandExecutor { player, args ->
            val recipe = args["recipe"] as Recipe
            player.inventory.addItem(recipe.result)
        })
        .register()
    // #endregion registerCommandWithSafeSuggestions

    // #region createForbiddenMobsList
    val forbiddenMobs = listOf<EntityType>(EntityType.ENDER_DRAGON, EntityType.WITHER)
    val allowedMobs = EntityType.entries.toMutableList()
    allowedMobs.removeAll(forbiddenMobs) // Now contains everything except ender dragon and wither
    // #endregion createForbiddenMobsList

    // #region createSafeArguments
    val safeArguments = listOf<Argument<*>>(
        EntityTypeArgument("mob").replaceSafeSuggestions(SafeSuggestions.suggest { info ->
            if (info.sender().isOp) {
                // All entity types
                EntityType.entries.toTypedArray()
            } else {
                // Only allowedMobs
                allowedMobs.toTypedArray()
            }
        })
    )
    // #endregion createSafeArguments

    // #region registerSpawnMobCommand
    CommandAPICommand("spawnmob")
        .withArguments(safeArguments)
        .executesPlayer(PlayerCommandExecutor { player, args ->
            val entityType = args["mob"] as EntityType
            player.world.spawnEntity(player.location, entityType)
        })
        .register()
    // #endregion registerSpawnMobCommand

    // #region createSafePotionEffectArguments
    val safeArgs = mutableListOf<Argument<*>>()
    safeArgs.add(EntitySelectorArgument.OnePlayer("target"))
    safeArgs.add(
        PotionEffectArgument("potioneffect").replaceSafeSuggestions(SafeSuggestions.suggest { info ->
            val target = info.previousArgs()["target"] as Player

            // Convert PotionEffect[] into PotionEffectType[]
            target.activePotionEffects.map { it.type }.toTypedArray()
        })
    )
    // #endregion createSafePotionEffectArguments

    // #region registerRemoveEffectCommand
    CommandAPICommand("removeeffect")
        .withArguments(safeArgs)
        .executesPlayer(PlayerCommandExecutor { _, args ->
            val target = args["target"] as Player
            val potionEffect = args["potioneffect"] as PotionEffectType
            target.removePotionEffect(potionEffect)
        })
        .register()
    // #endregion registerRemoveEffectCommand
}